File: cracking_interview/stacks_and_queues/min_stack.cpp
#include <iostream>
#include <stack>

template<typename T>
class MinStack {
public:
    void Push(T value) {
        Stack.push(value);
        if (MinValues.empty() || value <= MinValues.top()) {
            MinValues.push(value);
        }
    }

    T Top() const {
       return Stack.top();
    }

    void Pop() {
        auto value = Stack.top();
        Stack.pop();

        if (MinValues.top() == value) {
            MinValues.pop();
        }
    }

    T Min() const {
        return MinValues.top();
    }
private:
    std::stack<T> Stack;
    std::stack<T> MinValues;
};

int main() {
    MinStack<int> s;

    s.Push(3);
    std::cout << s.Min() << std::endl;

    s.Push(1);
    std::cout << s.Min() << std::endl;

    s.Pop();
    std::cout << s.Top() << std::endl;
    std::cout << s.Min() << std::endl;

    return 0;
}

File: leetcode/MinStack.java
import java.util.*;

public class MinStack {
    Stack<Integer> mins;
    Stack<Integer> objects;

    public MinStack() {
        mins = new Stack();
        objects = new Stack();

    }

    public void push(int x) {
        objects.push(x);

        if (mins.empty() || x <= mins.peek()) {
            mins.push(x);
        }
    }

    public void pop() {
        if (objects.pop().equals(mins.peek())) {
            mins.pop();
        }
    }

    public int top() {
        return objects.peek();
    }

    public int getMin() {
        return mins.peek();
    }

    public static void main(String[] args) {
        MinStack minStack = new MinStack();

        minStack.push(3);
        System.out.println(minStack.top());
        minStack.pop();
    }
}

